### ISSUE TITLE
`poetry show --solve` or enhance `--why`

### ISSUE THREAD

**Issue by ThatXliner** [python-poetry/poetry#6381]
- [X] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [X] I have searched the [documentation](https://python-poetry.org/docs/) and believe that my question is not covered.

## Feature Request

Related: https://discord.com/channels/487711540787675139/487711540787675143/1015462018767147030

I want to see what is indirectly constraining my dependency from updating, whether that is my Python version or some other package. An example of usage would be to see why a package isn't being updated to the latest version seen on PyPi.

A format like https://discord.com/channels/487711540787675139/487711540787675143/1015470135663939614 would be nice

`poetry show foo --solve`:
```
Project requires foo >=4.2,<69
bar adds constraint <=50, union is now >=4.2,<=50
baz adds constraint >49, union is now 50
Selecting version 50 of package foo because of unified constraint >49,50
```
or
`poetry show numpy --why --solve`:
```
Project requires transformers which requires numpy >= 1.21.0
Latest version of numpy (1.22.2) requires Python >=3.8,<4.0, incompatible with current version (Python >=3.7,<4.0)
numpy 1.21.5 works.
Selecting version 1.21.5 of numpy because it works lol
```

**Comment by dimbleby** [comment]
> I want to see what is indirectly constraining my dependency from updating

the easy way to answer that question at the moment is to add an explicit requirement that your package is at a recent version.

Then if that really was impossible, the failed solve will tell you why.

(sometimes the solve will succeed, but some other package will be downgraded instead.  eg if the latest versions of A and B are incompatible, it's pretty much a coin-flip which 'wins')

**Comment by ThatXliner** [comment]
> the easy way to answer that question at the moment is to add an explicit requirement that your dependency is at a recent version.

Yes, that's what I did. It's a hacky workaround. This is just a nice to have

**Comment by dimbleby** [comment]
The lock file doesn't store the history of why decisions were made, it just stores the solution that was found.  Really the only way to answer the question "why is version X of foo not allowed?" is to re-solve with a constraint requiring foo==X.  

So I wouldn't call this a hacky workaround: in essence it _is_ the solution.

You could have a dedicated command for "try bumping foo and tell me what went wrong" I suppose.  It doesn't much excite me but you never know, someone might get interested and work up an MR.  

If you care about this then much the most likely way for it to happen would be for you to be that someone and do it yourself...

**Comment by ThatXliner** [comment]
I know it would be re-solving the whole thing. Now I think of it, maybe this is better suited as a plugin.

**Comment by dimbleby** [comment]
Actually `poetry add foo@latest` is almost exactly what you want: that will try to add the latest version of `foo` and either succeed, or fail with a message telling you what the problem is.

The only way in which it is not quite right is if it does succeed then maybe you didn't want `foo` as a top-level dependency.  However: if you care about `foo` being latest and poetry wasn't otherwise giving you that: well, it turns out you do want `foo` as a top-level dependency...

**Comment by ThatXliner** [comment]
> However: if you care about `foo` being latest and poetry wasn't otherwise giving you that: well, it turns out you do want `foo` as a top-level dependency...

Makes sense. My situation was this: apparently, NumPy had a security vulnerability. I'm not using NumPy directly, and the vulnerability doesn't affect me, but I wanted to update it just in case something bad could happen.
